File: framework/lib/slider.dart
part of uxml;

/**
 * Manages location and size of thumb element inside a track.
 *
 * Unless the thumb element specifies min/max/width constraints,
 * the thumb will be scaled to layoutHeight/range.
 *
 * @author:ferhat@ (Ferhat Buyukkokten)
 */
class Slider extends ValueRangeControl {
  /** Orientation property definition */
  static PropertyDefinition orientationProperty;
  /** Number of tick marks property definition */
  static PropertyDefinition ticksProperty;
  /** Thumb property definition */
  static PropertyDefinition thumbProperty;
  /** Thumb pressed property definition */
  static PropertyDefinition thumbPressedProperty;
  /** ThumbSize property definition */
  static ElementDef sliderElementDef;

  bool _thumbPressed;
  num _thumbStartPos;
  num _moveStartValue;

  // Name of optional track part.
  static const String TRACK_PART_ID = "trackPart";

  Slider() : super() {
    _thumbPressed = false;
    _thumbStartPos = 0.0;
    _moveStartValue = 0.0;
  }

  /**
   * Sets or returns orientation.
   */
  int get orientation {
    return getProperty(orientationProperty);
  }

  set orientation(int value) {
    setProperty(orientationProperty, value);
  }

  /**
   * Sets or returns thumb element.
   */
  UIElement get thumb {
    return getProperty(thumbProperty);
  }

  set thumb(UIElement element) {
    setProperty(thumbProperty, element);
  }

  /**
   * Returns true if thumb is currently in pressed state.
   */
  bool get isPressed => getProperty(thumbPressedProperty);

  /**
   * Sets or returns number of tick marks on slider.
   */
  int get ticks => getProperty(ticksProperty);

  set ticks(int value) {
    setProperty(ticksProperty, value);
  }

  /** Overrides UIElement.onMeasure to return size of element. */
  bool onMeasure(num availableWidth, num availableHeight) {
    num maxWidth = 0.0;
    num maxHeight = 0.0;
    if (chromeTree != null) {
      chromeTree.measure(availableWidth, availableHeight);
      maxWidth = chromeTree.measuredWidth;
      maxHeight = chromeTree.measuredHeight;
    }

    UIElement thumbElement = thumb;
    if (thumbElement != null) {
      thumbElement.measure(availableWidth, availableHeight);
      if (thumbElement.measuredWidth > maxWidth) {
        maxWidth = thumbElement.measuredWidth;
      }
      if (thumbElement.measuredHeight > maxHeight) {
        maxHeight = thumbElement.measuredHeight;
      }
    }
    setMeasuredDimension(maxWidth, maxHeight);
    return false;
  }

  /** Overrides UIElement.onLayout. */
  void onLayout(num targetX, num targetY, num targetWidth, num targetHeight) {
    UIElement thumbElement = thumb;
    if (chromeTree != null) {
      chromeTree.layout(0.0, 0.0, targetWidth, targetHeight);
    }
    if (thumbElement == null) {
      return;
    }
    if (orientation == UIElement.VERTICAL) {
      thumbElement.opacity = (targetHeight >= thumbElement.measuredHeight) &&
          (thumbElement.measuredHeight != 0) ? 1.0 : 0.0;
      thumbElement.layout(0.0, (value - minValue) * _valueScaleFactor, max(
          thumbElement.measuredWidth, targetWidth), _thumbSize);
    } else {
      thumbElement.layout((value - minValue) * _valueScaleFactor, 0.0,
          _thumbSize, max(thumbElement.measuredHeight, targetHeight));
    }
  }

  /**
   * Returns scale factor for value to coordinate translation.
   */
  num get _valueScaleFactor {
    if ((maxValue - minValue).abs() < 0.000001) {
      return 0.0;
    }
    if (orientation == UIElement.VERTICAL) {
      return (layoutHeight - _thumbSize) / (maxValue - minValue);
    } else {
      return (layoutWidth - _thumbSize) / (maxValue - minValue);
    }
  }

  num get _thumbSize {
    // maxValue - minValue is the scroll range (contentSize - viewportSize).
    // to scale thumb, we need to compute visiblecontextsize/viewportsize.
    num s;
    if (orientation == UIElement.VERTICAL) {
      if (thumb.overridesProperty(UIElement.heightProperty) == false &&
          thumb.overridesProperty(UIElement.minHeightProperty)) {
        if ((layoutHeight + maxValue - minValue).abs() < 0.000001) {
          return layoutHeight;
        }
        s = max(thumb.minHeight, layoutHeight * (layoutHeight /
            (layoutHeight + maxValue - minValue)));
      } else {
        s = thumb.measuredHeight;
      }
    } else {
      if (thumb.overridesProperty(UIElement.widthProperty) == false &&
          thumb.overridesProperty(UIElement.minWidthProperty)) {
        if ((layoutWidth + maxValue - minValue).abs() < 0.000001) {
          return layoutHeight;
        }
        s = max(thumb.minWidth, layoutWidth * (layoutWidth /
            (layoutWidth + maxValue - minValue)));
      } else {
        s = thumb.measuredWidth;
      }
    }
    return s;
  }

  // Resize thumb when maxValue constraint changes.
  void _applyConstraints() {
    super._applyConstraints();
    if (thumb == null) {
      return;
    }
    if (orientation == UIElement.VERTICAL) {
      if (thumb.overridesProperty(UIElement.heightProperty) == false &&
          thumb.overridesProperty(UIElement.minHeightProperty)) {
        invalidateLayout();
      }
    } else if (thumb.overridesProperty(UIElement.widthProperty) == false &&
        thumb.overridesProperty(UIElement.minWidthProperty)) {
      invalidateLayout();
    }
  }

  static void _thumbChangedHandler(Object target,
                                   PropertyDefinition propDef,
                                   Object oldValue,
                                   Object newValue) {
    Slider slider = target;
    slider._thumbChanged(oldValue, newValue);
  }

  static void _ticksChangedHandler(Object target,
                                   PropertyDefinition propDef,
                                   Object oldValue,
                                   Object newValue) {
    if (newValue != 0) {
      Cursor.setElementCursor(target, Cursor.BUTTON);
    }
  }

  /** Override UIElement.getRawChildCount. */
  int getRawChildCount() {
    return ((chromeTree != null) ? 1 : 0) + ((thumb != null) ? 1: 0);
  }

  /** Override UIElement.getRawChild. */
  UIElement getRawChild(int index) {
    return ((index == 0) && (chromeTree != null)) ? chromeTree : thumb;
  }

  void _thumbChanged(UIElement prevThumb, UIElement newThumb) {
    if (prevThumb != null) {
      prevThumb.removeListener(UIElement.mouseDownEvent, _handleThumbMouseDown);
      removeRawChild(prevThumb);
    }
    if (newThumb != null) {
      insertRawChild(newThumb, -1);
      newThumb.addListener(UIElement.mouseDownEvent, _handleThumbMouseDown);
    }
  }

  void _handleThumbMouseDown(MouseEventArgs args) {
    captureMouse();
    _moveStartValue = value;
    _thumbStartPos = (orientation == UIElement.VERTICAL) ?
        args.getMousePosition(this).y : args.getMousePosition(this).x;
    _thumbPressed = true;
    args.handled = true;
  }

  /** Overrides UIElement.onMouseDown. */
  void onMouseDown(MouseEventArgs args) {
    UIElement trackElement = getElement(TRACK_PART_ID);
    if (trackElement != null) {
      num trackPos = 0.0;
      if (orientation == UIElement.VERTICAL) {
        num trackHeight = trackElement.layoutHeight;
        if (trackHeight == 0) {
          return;
        }
        trackPos = args.getMousePosition(trackElement).y / trackHeight;
        trackPos = (trackPos * (maxValue - minValue)) + minValue;
      } else {
        // horizontal slider
        num trackWidth = trackElement.layoutWidth;
        if (trackWidth == 0) {
          return;
        }
        trackPos = args.getMousePosition(trackElement).x / trackWidth;
        trackPos = (trackPos * (maxValue - minValue)) + minValue;
      }
      trackPos = normalizeTrackPos(trackPos);
      animate(ValueRangeControl.valueProperty, trackPos, duration: 150);
    }
    args.handled = true;
    captureMouse();
  }

  /**
   * Normalizes the track position by checking for min/max limits and
   * ticks setting.
   */
  num normalizeTrackPos(num trackPos) {
    if (ticks != 0) {
      trackPos = (trackPos - minValue) / (maxValue - minValue); // 0..1
      trackPos = ((((trackPos + (0.5 / ticks)) * ticks).toInt() / ticks) *
          (maxValue - minValue)) + minValue;
    }
    if (trackPos < minValue) {
      trackPos = minValue;
    }
    if (trackPos > maxValue) {
      trackPos = maxValue;
    }
    return trackPos;
  }

  /** Overrides UIElement.onMouseUp. */
  void onMouseUp(MouseEventArgs args) {
    if (ticks != 0) {
      num prevValue = value;
      num newValue = normalizeTrackPos(value);
      if (newValue != prevValue) {
        value = newValue;
      }
    }
    releaseMouse();
    _thumbPressed = false;
    args.handled = true;
  }

  /** Overrides UIElement.onMouseMove. */
  void onMouseMove(MouseEventArgs args) {
    if (_thumbPressed) {
      num newPos = (orientation == UIElement.VERTICAL) ?
          args.getMousePosition(this).y : args.getMousePosition(this).x;
      num newValue = _moveStartValue + ((newPos - _thumbStartPos)) /
          _valueScaleFactor;
      value = normalizeTrackPos(newValue);
      args.handled = true;
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => sliderElementDef;

  /** Registers component. */
  static void registerSlider() {
    orientationProperty = ElementRegistry.registerProperty("Orientation",
        PropertyType.ORIENTATION, PropertyFlags.RELAYOUT | PropertyFlags.RESIZE,
        null, UIElement.HORIZONTAL);
    ticksProperty = ElementRegistry.registerProperty("Ticks",
        PropertyType.INT, PropertyFlags.NONE, _ticksChangedHandler, 0);
    thumbProperty = ElementRegistry.registerProperty("Thumb",
        PropertyType.UIELEMENT, PropertyFlags.NONE, _thumbChangedHandler, null);
    thumbPressedProperty = ElementRegistry.registerProperty("isPressed",
        PropertyType.BOOL, PropertyFlags.NONE, null, false);
    sliderElementDef = ElementRegistry.register("Slider",
        ValueRangeControl.valuerangecontrolElementDef,
        [orientationProperty, ticksProperty, thumbProperty,
        thumbPressedProperty], null);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Slider.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.EventArgs;
import com.hello.uxml.tools.framework.events.EventHandler;
import com.hello.uxml.tools.framework.events.EventNotifier;
import com.hello.uxml.tools.framework.events.MouseEventArgs;

import java.util.EnumSet;

/**
 * Handles movement and sizing of a thumb inside a track.
 *
 * @author ferhat
 */
public class Slider extends ValueRangeControl {

  /** cached thumb control */
  protected UIElement cachedThumb;
  protected UIElement cachedTrack;
  private EventHandler thumbMouseDownHandler;
  private double moveStartValue;
  private double thumbStartPos;

  /** Thumb Property Definition */
  public static PropertyDefinition thumbPropDef = PropertySystem.register("Thumb",
      UIElement.class, Slider.class,
      new PropertyData(null, EnumSet.of(PropertyFlags.Resize), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          Slider slider = (Slider) e.getSource();
          slider.thumbChanged((UIElement) e.getOldValue(), (UIElement) e.getNewValue());
        }
      }));

  /** Orientation Property Definition */
  public static PropertyDefinition orientationPropDef = PropertySystem.register("Orientation",
      Orientation.class, Slider.class,
      new PropertyData(Orientation.Horizontal, EnumSet.of(PropertyFlags.Resize)));

  /** Ticks Property Definition */
  public static PropertyDefinition ticksPropDef = PropertySystem.register("Ticks",
      Integer.class, Slider.class,
      new PropertyData(0));

  /** ThumbPressed property definition */
  public static PropertyDefinition thumbPressedPropDef = PropertySystem.register("ThumbPressed",
      Boolean.class, Slider.class,
      new PropertyData(false, EnumSet.of(PropertyFlags.None)));

  // Name of optional track part.
  private static final String TRACK_PART_ID = "trackPart";

  /**
   * Sets or returns thumb element of slider.
   */
  public void setThumb(UIElement content) {
    setProperty(thumbPropDef, content);
  }

  public UIElement getThumb() {
    return cachedThumb;
  }

  /**
   * Sets or returns tick mark count.
   */
  public void setTicks(int value) {
    setProperty(ticksPropDef, value);
  }

  public int getTicks() {
    return ((Integer) getProperty(ticksPropDef)).intValue();
  }

  /**
   * Returns whether thumb is in pressed down state.
   */
  public boolean getThumbPressed() {
    return (Boolean) getProperty(thumbPressedPropDef);
  }

  private void setThumbPressed(boolean val) {
    setProperty(thumbPressedPropDef, val);
  }


  /**
   * Sets or returns orientation.
   */
  public void setOrientation(Orientation orientation) {
    setProperty(orientationPropDef, orientation);
  }

  public Orientation getOrientation() {
    return (Orientation) getProperty(orientationPropDef);
  }

  @Override
  protected void onMeasure(double availableWidth, double availableHeight) {
    double maxWidth = 0;
    double maxHeight = 0;
    if (chromeTree != null) {
      chromeTree.measure(availableWidth, availableHeight);
      maxWidth = chromeTree.getMeasuredWidth();
      maxHeight = chromeTree.getMeasuredHeight();
    }
    UIElement thumbElement = getThumb();
    if (thumbElement != null) {
      thumbElement.measure(availableWidth, availableHeight);
      if (thumbElement.getMeasuredWidth() > maxWidth) {
        maxWidth = thumbElement.getMeasuredWidth();
      }
      if (thumbElement.getMeasuredHeight() > maxHeight) {
        maxHeight = thumbElement.getMeasuredHeight();
      }
    }
    setMeasuredDimension(maxWidth, maxHeight);
  }

  /**
   * Returns scale factor for value to coordinate translation.
   */
  private double getValueScaleFactor() {
    if (getOrientation() == Orientation.Vertical) {
      return (getLayoutRect().height - getThumb().getMeasuredHeight()) /
      (getMaxValue() - getMinValue());
    } else {
      return (getLayoutRect().width - getThumb().getMeasuredWidth()) /
      (getMaxValue() - getMinValue());
    }
  }

  @Override
  protected void onLayout(Rectangle targetRect) {
    if (chromeTree != null) {
      chromeTree.layout(targetRect);
    }
    UIElement thumbElement = getThumb();
    if (thumbElement == null) {
      return;
    }
    if (getOrientation() == Orientation.Vertical) {
      thumbElement.layout(0, getValue() * getValueScaleFactor(),
          Math.max(thumbElement.getMeasuredWidth(), targetRect.width),
          thumbElement.getMeasuredHeight());
    } else {
      thumbElement.layout(getValue() * getValueScaleFactor(), 0, thumbElement.getMeasuredWidth(),
          Math.max(thumbElement.getMeasuredHeight(), targetRect.height));
    }
  }

  private void thumbChanged(UIElement prevThumb, UIElement newThumb) {
    if (cachedThumb != null) {
      cachedThumb.removeListener(UIElement.mouseDownEvent, thumbMouseDownHandler);
      removeRawChild(cachedThumb);
    }
    if (newThumb != null) {
      cachedThumb = newThumb;
      addRawChild(cachedThumb);
      if (thumbMouseDownHandler == null) {
        thumbMouseDownHandler = new EventHandler(){
          @Override
          public void handleEvent(EventNotifier targetObject, EventArgs e) {
            handleThumbMouseDown((MouseEventArgs) e);
          }};
      }
      cachedThumb.addListener(UIElement.mouseDownEvent, thumbMouseDownHandler);
    } else {
      cachedThumb = null;
    }
  }

  @Override
  protected int getRawChildCount() {
    return ((chromeTree != null) ? 1 : 0) + ((cachedThumb != null) ? 1 : 0);
  }

  @Override
  protected UIElement getRawChild(int index) {
    return ((index == 0)  && (chromeTree != null)) ? chromeTree : cachedThumb;
  }

  private void handleThumbMouseDown(MouseEventArgs e) {
    captureMouse();
    moveStartValue = getValue();
    thumbStartPos = (getOrientation() == Orientation.Vertical)
        ? e.getMousePosition(this).y
        : e.getMousePosition(this).x;
    setThumbPressed(true);
    e.setHandled(true);
  }

  @Override
  protected void onMouseMove(MouseEventArgs e) {
    if (!getThumbPressed()) {
      return;
    }
    double newPos =  (getOrientation() == Orientation.Vertical)
        ? e.getMousePosition(this).y
        : e.getMousePosition(this).x;
    double newValue = moveStartValue + ((newPos - thumbStartPos)) / getValueScaleFactor();
    setValue(normalizeTrackPos(newValue));
    e.setHandled(true);
  }

  private double normalizeTrackPos(double trackPos) {
    int ticks = getTicks();
    double minValue = getMinValue();
    double maxValue = getMaxValue();
    if (ticks != 0) {
      trackPos = (trackPos - minValue) / (maxValue - minValue); // 0..1
      trackPos = (((int) ((trackPos + (0.5 / ticks)) * ticks) / ticks) *
        (maxValue - minValue)) + minValue;
    }
    trackPos = Math.min(Math.max(trackPos, minValue), maxValue);
    return trackPos;
  }

  // Handles slider position change when user clicks directly on
  // slider track instead of dragging thumb.
  @Override
  protected void onMouseDown(MouseEventArgs args) {
    UIElement trackElement = getElement(TRACK_PART_ID);
    if (trackElement != null) {
      double trackPos = 0;
      double minValue = getMinValue();
      double maxValue = getMaxValue();
      if (getOrientation() == Orientation.Vertical) {
        double trackHeight = trackElement.getLayoutRect().height;
        if (trackHeight == 0) {
          return;
        }
        trackPos = args.getMousePosition(trackElement).y / trackHeight;
        trackPos = (trackPos * (maxValue - minValue)) + minValue;
      } else {
        // horizontal slider
        double trackWidth = trackElement.getLayoutRect().width;
        if (trackWidth == 0) {
          return;
        }
        trackPos = args.getMousePosition(trackElement).x / trackWidth;
        trackPos = (trackPos * (maxValue - minValue)) + minValue;
      }
      trackPos = normalizeTrackPos(trackPos);
      setValue(trackPos);
    }
    args.setHandled(true);
  }

  @Override
  protected void onMouseUp(MouseEventArgs e) {
    if (getTicks() != 0) {
      double prevValue = getValue();
      double newValue = normalizeTrackPos(prevValue);
      if (newValue != prevValue) {
        setValue(newValue);
      }
    }
    releaseMouse();
    setThumbPressed(false);
    e.setHandled(true);
  }
}
